File: backend/hackathon/src/main/java/com/dev_trails/hackathon/InsuranceService.java
package com.dev_trails.hackathon;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;
import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.*;

@Service
public class InsuranceService {
    private final RiderRepository riderRepo;
    private final PayoutLogRepository payoutLogRepo;
    private final PolicyRepository policyRepo;
    private final ClaimRepository claimRepo;
    private final FraudLogRepository fraudLogRepo;
    private final ReferralRepository referralRepo;
    private final RestTemplate restTemplate;

    @Value("${oracle.base-url}")
    private String oracleBaseUrl;

    // Market crash state — volatile for thread safety
    private volatile boolean marketCrashActive = false;
    private volatile String marketCrashAction = "NONE";

    public InsuranceService(RiderRepository riderRepo, PayoutLogRepository payoutLogRepo,
                            PolicyRepository policyRepo, ClaimRepository claimRepo,
                            FraudLogRepository fraudLogRepo, ReferralRepository referralRepo,
                            RestTemplate restTemplate) {
        this.riderRepo = riderRepo;
        this.payoutLogRepo = payoutLogRepo;
        this.policyRepo = policyRepo;
        this.claimRepo = claimRepo;
        this.fraudLogRepo = fraudLogRepo;
        this.referralRepo = referralRepo;
        this.restTemplate = restTemplate;
    }

    // ── Rider Registration ──────────────────────────────────────────────────

    public Rider registerRider(String name, String phone, String city, String zone, String platform, Integer age) {
        Rider r = new Rider();
        r.name = name;
        r.phone = phone;
        r.city = city;
        r.zone = zone;
        r.platform = platform;
        r.age = age;
        r.walletBalance = 500.0;
        r.isPolicyActive = false;
        r.policyTier = null;
        r.registeredAt = LocalDate.now().toString();
        // Generate unique referral code
        r.referralCode = "RW-" + phone.substring(Math.max(0, phone.length() - 4)) + "-" + System.currentTimeMillis() % 10000;
        return riderRepo.save(r);
    }

    public Rider getRider(Long riderId) {
        return riderRepo.findById(riderId).orElseThrow(() -> new RuntimeException("Rider not found: " + riderId));
    }

    public Rider updateRider(Long riderId, Rider updates) {
        Rider r = getRider(riderId);
        if (updates.name != null) r.name = updates.name;
        if (updates.phone != null) r.phone = updates.phone;
        if (updates.city != null) r.city = updates.city;
        if (updates.platform != null) r.platform = updates.platform;
        if (updates.age != null) r.age = updates.age;
        return riderRepo.save(r);
    }

    public List<PayoutLog> getPayouts(Long riderId) {
        return payoutLogRepo.findByRiderIdOrderByTimestampDesc(riderId);
    }

    // ── Quote & Policy Purchase ─────────────────────────────────────────────

    public Map<String, OracleForecastResponse.PlanDetail> getQuote(Long riderId) {
        Rider r = getRider(riderId);
        try {
            OracleForecastResponse res = restTemplate.getForObject(
                oracleBaseUrl + "/api/v1/pricing/forecast-quote?zone=" + r.zone,
                OracleForecastResponse.class);
            return res.plans;
        } catch (Exception e) {
            // Fallback pricing if Oracle is unreachable
            Map<String, OracleForecastResponse.PlanDetail> fallback = new LinkedHashMap<>();
            OracleForecastResponse.PlanDetail basic = new OracleForecastResponse.PlanDetail();
            basic.premium = 95.0; basic.daily_payout = 300.0;
            OracleForecastResponse.PlanDetail standard = new OracleForecastResponse.PlanDetail();
            standard.premium = 143.0; standard.daily_payout = 500.0;
            OracleForecastResponse.PlanDetail pro = new OracleForecastResponse.PlanDetail();
            pro.premium = 218.0; pro.daily_payout = 1000.0;
            fallback.put("basic", basic);
            fallback.put("standard", standard);
            fallback.put("pro", pro);
            return fallback;
        }
    }

    public Rider buyPolicy(Long riderId, String tier) {
        Rider r = getRider(riderId);

        // ── Dynamic Solvency Protocol: Block Pro/Standard during market crash ──
        if (marketCrashActive && ("PRO".equalsIgnoreCase(tier) || "STANDARD".equalsIgnoreCase(tier))) {
            throw new RuntimeException("MARKET_CRASH_PROTOCOL: " + tier.toUpperCase() +
                    " tier is temporarily locked. Only Basic tier is available during market crisis.");
        }

        double premium = 0;
        String planName = "";
        try {
            OracleForecastResponse res = restTemplate.getForObject(
                oracleBaseUrl + "/api/v1/pricing/forecast-quote?zone=" + r.zone,
                OracleForecastResponse.class);
            OracleForecastResponse.PlanDetail plan = res.plans.get(tier.toLowerCase());
            if (plan != null) premium = plan.premium;
        } catch (Exception e) {
            premium = switch (tier.toUpperCase()) {
                case "PRO" -> 218.0;
                case "STANDARD" -> 143.0;
                default -> 95.0;
            };
        }
        planName = switch (tier.toUpperCase()) {
            case "PRO" -> "Heat Shield Pro";
            case "STANDARD" -> "Rain Guard Plus";
            default -> "Heat Shield Basic";
        };
        r.walletBalance -= premium;
        r.isPolicyActive = true;
        r.policyTier = tier.toUpperCase();
        Rider saved = riderRepo.save(r);

        // Create a Policy record so the admin PolicyCenter can see this user
        Policy p = new Policy();
        p.policyNumber = "POL-GW-" + LocalDate.now().getYear() + "-" + String.format("%03d", saved.id);
        p.riderId = saved.id;
        p.riderName = saved.name;
        p.zone = saved.zone + " (" + saved.city + ")";
        p.plan = planName;
        p.premium = "\u20B9" + (int) premium + "/day";
        p.riskScore = 50;
        p.status = "Active";
        p.startDate = LocalDate.now().toString();
        p.phone = saved.phone;
        p.email = saved.name.toLowerCase().replaceAll("\\s+", ".") + "@gmail.com";
        p.address = saved.city;
        p.birthdate = "2000-01-01";
        p.customerSince = LocalDate.now().toString();
        p.accountTier = tier.toUpperCase().equals("PRO") ? "Gold" : (tier.toUpperCase().equals("STANDARD") ? "Silver" : "Bronze");
        p.delinquencyStatus = "NO";
        policyRepo.save(p);

        return saved;
    }

    // ── Actuarial Engine with Dual Validation + Fraud Check ──────────────

    @Scheduled(cron = "0 0 * * * *")
    public void runActuarialEngine() {
        List<String> zones = riderRepo.findDistinctActiveZones();
        for (String zone : zones) {
            try {
                // ── Layer 1: External Weather Disruption Check ──
                OracleLiveResponse live = restTemplate.getForObject(
                    oracleBaseUrl + "/api/v1/pricing/quote?zone=" + zone,
                    OracleLiveResponse.class);
                if (live == null || !Boolean.TRUE.equals(live.payout_triggered)) continue;

                // ── Layer 2: Platform Status (Dual Validation) ──
                // Verify platform order volume has actually dropped
                boolean platformDisrupted = true;
                try {
                    Map platformStatus = restTemplate.getForObject(
                        oracleBaseUrl + "/api/v1/oracle/platform-status/" + zone,
                        Map.class);
                    if (platformStatus != null) {
                        Object confirmed = platformStatus.get("disruption_confirmed");
                        platformDisrupted = Boolean.TRUE.equals(confirmed);
                    }
                } catch (Exception e) {
                    // If platform check fails, rely on weather trigger alone
                    platformDisrupted = true;
                }

                if (!platformDisrupted) continue;  // Dual validation failed

                List<Rider> activeRiders = riderRepo.findByZoneAndIsPolicyActiveTrue(zone);
                for (Rider rider : activeRiders) {
                    // ── Layer 3: Fraud Verification ──
                    boolean isFraud = false;
                    try {
                        Map<String, Object> fraudPayload = new HashMap<>();
                        fraudPayload.put("claim_id", "CLM-AUTO-" + rider.id + "-" + System.currentTimeMillis());
                        fraudPayload.put("gps_lat", 13.08);
                        fraudPayload.put("gps_lon", 80.27);
                        fraudPayload.put("is_mock_flag", false);
                        fraudPayload.put("network_rtt_ms", 35.0);
                        fraudPayload.put("device_speed", 5.0);
                        fraudPayload.put("altitude", 15.0);
                        fraudPayload.put("asn_type", "mobile");

                        FraudCheckResponse fraudCheck = restTemplate.postForObject(
                            oracleBaseUrl + "/api/v1/oracle/verify-claim",
                            fraudPayload, FraudCheckResponse.class);

                        if (fraudCheck != null) {
                            isFraud = Boolean.TRUE.equals(fraudCheck.fraud_flag);

                            // Log fraud check result
                            FraudLog log = new FraudLog();
                            log.claimId = fraudCheck.claim_id;
                            log.riderId = rider.id;
                            log.riderName = rider.name;
                            log.fraudFlag = fraudCheck.fraud_flag;
                            log.confidenceScore = fraudCheck.confidence_score;
                            log.fraudScore = fraudCheck.fraud_score;
                            log.mlPrediction = fraudCheck.ml_prediction;
                            log.fraudReasons = fraudCheck.fraud_reasons != null ?
                                String.join(",", fraudCheck.fraud_reasons) : "";
                            log.zone = zone;
                            log.gpsLat = 13.08;
                            log.gpsLon = 80.27;
                            log.networkRttMs = 35.0;
                            log.timestamp = LocalDateTime.now().toString();
                            log.verdict = isFraud ? "BLOCKED" : "PASSED";
                            fraudLogRepo.save(log);
                        }
                    } catch (Exception e) {
                        // If fraud check fails, proceed with payout (fail-open for rider safety)
                    }

                    if (isFraud) continue;  // Skip payout for fraudulent claims

                    // ── Execute Payout ──
                    double amount = switch (rider.policyTier) {
                        case "STANDARD" -> 500.0;
                        case "PRO" -> 1000.0;
                        default -> 300.0;
                    };

                    // During market crash, ensure Basic tier always pays out (Zero-Delay Guarantee)
                    if (marketCrashActive && !"BASIC".equals(rider.policyTier)) {
                        amount = 300.0;  // Reduce to basic payout during crash
                    }

                    rider.walletBalance += amount;
                    rider.isPolicyActive = false;
                    riderRepo.save(rider);

                    PayoutLog payoutLog = new PayoutLog();
                    payoutLog.riderId = rider.id;
                    payoutLog.amount = amount;
                    payoutLog.timestamp = LocalDateTime.now();
                    payoutLog.triggerType = "WEATHER_PARAMETRIC";
                    payoutLog.zone = zone;
                    payoutLog.claimNumber = "CLM-" + System.currentTimeMillis() + "-" + rider.id;
                    payoutLogRepo.save(payoutLog);

                    // Create claim record
                    Claim claim = new Claim();
                    claim.claimNumber = payoutLog.claimNumber;
                    claim.policyRef = "POL-GW-" + LocalDate.now().getYear() + "-" + String.format("%03d", rider.id);
                    claim.riderId = rider.id;
                    claim.riderName = rider.name;
                    claim.product = rider.policyTier;
                    claim.fraudRisk = "NO";
                    claim.dateOfLoss = LocalDate.now().toString();
                    claim.status = "AUTO-APPROVED";
                    claim.triggerType = "WEATHER";
                    claim.amount = amount;
                    claim.zone = zone;
                    claim.approvedAt = LocalDateTime.now().toString();
                    claimRepo.save(claim);
                }
            } catch (Exception e) {
                // Zone processing failed, continue to next zone
            }
        }
    }

    // ── Market Crash Monitor (Dynamic Solvency Protocol) ────────────────

    @Scheduled(cron = "0 30 * * * *")  // Every 30 minutes
    public void runMarketCrashMonitor() {
        String[] cities = {"chennai", "delhi", "mumbai", "bangalore", "hyderabad"};
        for (String city : cities) {
            try {
                MarketHealthResponse health = restTemplate.getForObject(
                    oracleBaseUrl + "/api/v1/oracle/market-health/" + city,
                    MarketHealthResponse.class);
                if (health != null && Boolean.TRUE.equals(health.crash_detected)) {
                    marketCrashActive = true;
                    marketCrashAction = "LOCK_PRO_STANDARD_TIERS";
                    return;  // One crash triggers protocol globally
                }
            } catch (Exception e) {
                // Continue checking other cities
            }
        }
        // If no crash detected, clear protocol
        marketCrashActive = false;
        marketCrashAction = "NONE";
    }

    // ── Claim Tracker (Zero-Touch Pipeline Status) ──────────────────────

    public Map<String, Object> getClaimTracker(Long riderId) {
        Rider r = getRider(riderId);
        Map<String, Object> tracker = new LinkedHashMap<>();

        // Determine pipeline stage
        String stage = "MONITORING";
        boolean triggered = false;
        boolean validated = false;
        boolean paid = false;

        // Check if there's a recent weather trigger for this zone
        try {
            OracleLiveResponse live = restTemplate.getForObject(
                oracleBaseUrl + "/api/v1/pricing/quote?zone=" + r.zone,
                OracleLiveResponse.class);
            if (live != null && Boolean.TRUE.equals(live.payout_triggered)) {
                triggered = true;
                stage = "TRIGGER_MET";
                tracker.put("trigger_type", live.trigger_type != null ? live.trigger_type : "WEATHER");
                tracker.put("weather", Map.of(
                    "temp", live.live_temp != null ? live.live_temp : 30.0,
                    "rain", live.live_rain != null ? live.live_rain : 0.0
                ));
            }
        } catch (Exception e) {
            // Oracle unreachable
        }

        // Check recent payouts
        List<PayoutLog> payouts = payoutLogRepo.findByRiderIdOrderByTimestampDesc(riderId);
        if (!payouts.isEmpty()) {
            PayoutLog latest = payouts.get(0);
            // If payout was within last 24 hours, show completed pipeline
            if (latest.timestamp != null && latest.timestamp.isAfter(LocalDateTime.now().minusHours(24))) {
                paid = true;
                validated = true;
                triggered = true;
                stage = "PAYOUT_SENT";
                tracker.put("payout_amount", latest.amount);
                tracker.put("payout_time", latest.timestamp.toString());
                tracker.put("claim_number", latest.claimNumber);
            }
        }

        // Check validating status (claim exists but not yet approved)
        List<Claim> pendingClaims = claimRepo.findByRiderId(riderId);
        for (Claim c : pendingClaims) {
            if ("Open".equals(c.status)) {
                validated = false;
                triggered = true;
                stage = "VALIDATING_SHIFT";
            }
        }

        tracker.put("rider_id", riderId);
        tracker.put("zone", r.zone);
        tracker.put("policy_active", r.isPolicyActive);
        tracker.put("current_stage", stage);
        tracker.put("pipeline", Map.of(
            "monitoring_weather", true,
            "trigger_met", triggered,
            "validating_shift", validated || (triggered && stage.equals("VALIDATING_SHIFT")),
            "payout_sent", paid
        ));

        return tracker;
    }

    // ── Market Status ────────────────────────────────────────────────────

    public Map<String, Object> getMarketStatus() {
        Map<String, Object> status = new LinkedHashMap<>();
        status.put("crash_active", marketCrashActive);
        status.put("action", marketCrashAction);
        status.put("pro_tier_locked", marketCrashActive);
        status.put("standard_tier_locked", marketCrashActive);
        status.put("basic_tier_available", true);
        status.put("timestamp", LocalDateTime.now().toString());
        return status;
    }

    // ── Referral System ─────────────────────────────────────────────────

    public Map<String, Object> generateReferralCode(Long riderId) {
        Rider r = getRider(riderId);
        // Check if rider already has a referral record
        Optional<ReferralCode> existing = referralRepo.findByOwnerRiderId(riderId);
        if (existing.isPresent()) {
            ReferralCode ref = existing.get();
            return Map.of(
                "code", ref.code,
                "times_used", ref.timesUsed,
                "reward_earned", ref.rewardEarned,
                "share_message", "Join RiskWire with my code " + ref.code + " and get ₹50 bonus! Download: https://riskwire.app"
            );
        }
        // Generate new referral
        String code = r.referralCode != null ? r.referralCode : "RW-" + riderId + "-" + System.currentTimeMillis() % 10000;
        ReferralCode ref = new ReferralCode();
        ref.code = code;
        ref.ownerRiderId = riderId;
        ref.ownerName = r.name;
        ref.timesUsed = 0;
        ref.rewardEarned = 0.0;
        ref.createdAt = LocalDate.now().toString();
        referralRepo.save(ref);

        return Map.of(
            "code", code,
            "times_used", 0,
            "reward_earned", 0.0,
            "share_message", "Join RiskWire with my code " + code + " and get ₹50 bonus! Download: https://riskwire.app"
        );
    }

    public Map<String, Object> redeemReferral(Long riderId, String code) {
        Optional<ReferralCode> refOpt = referralRepo.findByCode(code);
        if (refOpt.isEmpty()) {
            throw new RuntimeException("Invalid referral code: " + code);
        }
        ReferralCode ref = refOpt.get();
        if (ref.ownerRiderId.equals(riderId)) {
            throw new RuntimeException("Cannot use your own referral code");
        }

        // Reward both parties
        Rider newUser = getRider(riderId);
        newUser.walletBalance += 50.0;  // ₹50 bonus for new user
        newUser.referredBy = code;
        riderRepo.save(newUser);

        Rider owner = getRider(ref.ownerRiderId);
        owner.walletBalance += 25.0;  // ₹25 for referrer
        riderRepo.save(owner);

        ref.timesUsed += 1;
        ref.rewardEarned += 25.0;
        referralRepo.save(ref);

        return Map.of(
            "success", true,
            "new_user_bonus", 50.0,
            "referrer_bonus", 25.0,
            "message", "₹50 added to your wallet!"
        );
    }
}

File: backend/services/insuranceService.js
const axios = require('axios');
const { Rider, PayoutLog, Policy, Claim, FraudLog, ReferralCode, BillingTransaction } = require('../models');

const ORACLE_BASE_URL = process.env.ORACLE_BASE_URL;

// Market crash state
let marketCrashActive = false;
let marketCrashAction = 'NONE';

// ── Getter/Setter for market crash state ──────────────────────────────────
function isMarketCrashActive() { return marketCrashActive; }
function setMarketCrashState(active, action) {
  marketCrashActive = active;
  marketCrashAction = action;
}

// ── Rider Registration ────────────────────────────────────────────────────

async function registerRider(name, phone, city, zone, platform, age, workerId) {
  const rider = await Rider.create({
    name,
    phone,
    city,
    zone,
    platform,
    age: age || 25,
    wallet_balance: 500.0,
    is_policy_active: false,
    policy_tier: null,
    registered_at: new Date().toISOString().split('T')[0],
    worker_id: workerId || '',
    verified: !!(workerId && workerId.length > 0),
    referral_code: 'RW-' + phone.substring(Math.max(0, phone.length - 4)) + '-' + (Date.now() % 10000),
  });
  return formatRider(rider);
}

async function getRider(riderId) {
  // Support both Mongo ObjectId strings and legacy numeric IDs
  const rider = await Rider.findById(riderId).catch(() => null)
    || await Rider.findOne({ phone: riderId }).catch(() => null);
  if (!rider) throw new Error('Rider not found: ' + riderId);
  return rider;
}

async function updateRider(riderId, updates) {
  const rider = await getRider(riderId);
  if (updates.name != null) rider.name = updates.name;
  if (updates.phone != null) rider.phone = updates.phone;
  if (updates.city != null) rider.city = updates.city;
  if (updates.platform != null) rider.platform = updates.platform;
  if (updates.age != null) rider.age = updates.age;
  await rider.save();
  return formatRider(rider);
}

async function getPayouts(riderId) {
  const payouts = await PayoutLog.find({ rider_id: riderId }).sort({ timestamp: -1 });
  return payouts.map(formatPayoutLog);
}

// ── Quote & Policy Purchase ───────────────────────────────────────────────

async function getQuote(riderId) {
  const rider = await getRider(riderId);
  try {
    const res = await axios.get(`${ORACLE_BASE_URL}/api/v1/pricing/forecast-quote?zone=${rider.zone}`);
    return res.data.plans;
  } catch (e) {
    return {
      basic: { premium: 95.0, daily_payout: 300.0 },
      standard: { premium: 143.0, daily_payout: 500.0 },
      pro: { premium: 218.0, daily_payout: 1000.0 },
    };
  }
}

async function buyPolicy(riderId, tier) {
  const rider = await getRider(riderId);

  // ── Item 8: Adverse Selection Lockout ─────────────────────────────────
  // Block new enrollments if a weather red-alert is already active in the rider's zone.
  // This prevents riders from buying coverage only when a storm is imminent (48h rule).
  try {
    const alertRes = await axios.get(`${ORACLE_BASE_URL}/api/v1/pricing/quote?zone=${rider.zone}`);
    const live = alertRes.data;
    if (live && live.payout_triggered === true) {
      throw new Error(
        'ADVERSE_SELECTION_LOCKOUT: A weather red-alert is currently active in zone ' + rider.zone +
        '. New policy enrollment is locked 48 hours before and during active weather events. ' +
        'Please try again after the alert clears.'
      );
    }
  } catch (e) {
    // Only rethrow if it is our lockout error, not an oracle connectivity issue
    if (e.message && e.message.startsWith('ADVERSE_SELECTION_LOCKOUT')) throw e;
  }

  // ── Dynamic Solvency Protocol: Block Pro/Standard during market crash ──
  if (marketCrashActive && (tier.toUpperCase() === 'PRO' || tier.toUpperCase() === 'STANDARD')) {
    throw new Error(
      'MARKET_CRASH_PROTOCOL: ' + tier.toUpperCase() +
      ' tier is temporarily locked. Only Basic tier is available during market crisis.'
    );
  }

  let premium = 0;
  let planName = '';

  try {
    const res = await axios.get(`${ORACLE_BASE_URL}/api/v1/pricing/forecast-quote?zone=${rider.zone}`);
    const plan = res.data.plans[tier.toLowerCase()];
    if (plan) premium = plan.premium;
  } catch (e) {
    switch (tier.toUpperCase()) {
      case 'PRO': premium = 218.0; break;
      case 'STANDARD': premium = 143.0; break;
      default: premium = 95.0;
    }
  }

  switch (tier.toUpperCase()) {
    case 'PRO': planName = 'Heat Shield Pro'; break;
    case 'STANDARD': planName = 'Rain Guard Plus'; break;
    default: planName = 'Heat Shield Basic';
  }

  rider.wallet_balance -= premium;
  rider.is_policy_active = true;
  rider.policy_tier = tier.toUpperCase();
  await rider.save();

  const year = new Date().getFullYear();
  const policyNumber = `POL-GW-${year}-${String(rider.id).slice(-6)}`;
  const today = new Date().toISOString().split('T')[0];

  await Policy.create({
    policy_number: policyNumber,
    rider_id: rider.id,
    rider_name: rider.name,
    zone: rider.zone + ' (' + rider.city + ')',
    plan: planName,
    premium: '₹' + Math.floor(premium) + '/day',
    risk_score: 50,
    status: 'Active',
    start_date: today,
    phone: rider.phone,
    email: rider.name.toLowerCase().replace(/\s+/g, '.') + '@gmail.com',
    address: rider.city,
    birthdate: '2000-01-01',
    customer_since: today,
    account_tier: tier.toUpperCase() === 'PRO' ? 'Gold' : (tier.toUpperCase() === 'STANDARD' ? 'Silver' : 'Bronze'),
    delinquency_status: 'CURRENT',
  });

  await BillingTransaction.create({
    txn_id: 'TXN-PREM-' + Date.now() + '-' + rider.id,
    type: 'PREMIUM',
    rider_name: rider.name,
    amount: premium,
    date: today,
    description: `${planName} weekly premium - ${rider.city}`,
    policy_ref: policyNumber,
  });

  return formatRider(rider);
}

// ── Claim Tracker ────────────────────────────────────────────────────────

async function getClaimTracker(riderId) {
  const rider = await getRider(riderId);
  const tracker = {};

  let stage = 'MONITORING';
  let triggered = false;
  let validated = false;
  let paid = false;

  try {
    const res = await axios.get(`${ORACLE_BASE_URL}/api/v1/pricing/quote?zone=${rider.zone}`);
    const live = res.data;
    if (live && live.payout_triggered === true) {
      triggered = true;
      stage = 'TRIGGER_MET';
      tracker.trigger_type = live.trigger_type || 'WEATHER';
      tracker.weather = { temp: live.live_temp || 30.0, rain: live.live_rain || 0.0 };
    }
  } catch (e) { /* Oracle unreachable */ }

  const payouts = await PayoutLog.find({ rider_id: riderId }).sort({ timestamp: -1 });

  if (payouts.length > 0) {
    const latest = payouts[0];
    const twentyFourHoursAgo = new Date(Date.now() - 24 * 60 * 60 * 1000);
    if (latest.timestamp && new Date(latest.timestamp) > twentyFourHoursAgo) {
      paid = true; validated = true; triggered = true;
      stage = 'PAYOUT_SENT';
      tracker.payout_amount = latest.amount;
      tracker.payout_time = latest.timestamp instanceof Date
        ? latest.timestamp.toISOString()
        : String(latest.timestamp);
      tracker.claim_number = latest.claim_number;
    }
  }

  const pendingClaims = await Claim.find({ rider_id: riderId });
  for (const c of pendingClaims) {
    if (c.status === 'Open') {
      validated = false; triggered = true;
      stage = 'VALIDATING_SHIFT';
    }
  }

  tracker.rider_id = riderId;
  tracker.zone = rider.zone;
  tracker.policy_active = rider.is_policy_active;
  tracker.current_stage = stage;
  tracker.pipeline = {
    monitoring_weather: true,
    trigger_met: triggered,
    validating_shift: validated || (triggered && stage === 'VALIDATING_SHIFT'),
    payout_sent: paid,
  };

  return tracker;
}

// ── Market Status ─────────────────────────────────────────────────────────

function getMarketStatus() {
  return {
    crash_active: marketCrashActive,
    action: marketCrashAction,
    pro_tier_locked: marketCrashActive,
    standard_tier_locked: marketCrashActive,
    basic_tier_available: true,
    timestamp: new Date().toISOString(),
  };
}

// ── Referral System ───────────────────────────────────────────────────────

async function generateReferralCode(riderId) {
  const rider = await getRider(riderId);

  const existing = await ReferralCode.findOne({ owner_rider_id: riderId });
  if (existing) {
    return {
      code: existing.code,
      times_used: existing.times_used,
      reward_earned: existing.reward_earned,
      share_message: 'Join RiskWire with my code ' + existing.code + ' and get ₹50 bonus! Download: https://riskwire.app',
    };
  }

  const code = rider.referral_code || ('RW-' + rider.id.toString().slice(-4) + '-' + (Date.now() % 10000));
  await ReferralCode.create({
    code,
    owner_rider_id: riderId,
    owner_name: rider.name,
    times_used: 0,
    reward_earned: 0.0,
    created_at: new Date().toISOString().split('T')[0],
  });

  return {
    code,
    times_used: 0,
    reward_earned: 0.0,
    share_message: 'Join RiskWire with my code ' + code + ' and get ₹50 bonus! Download: https://riskwire.app',
  };
}

async function redeemReferral(riderId, code) {
  const ref = await ReferralCode.findOne({ code });
  if (!ref) throw new Error('Invalid referral code: ' + code);
  if (String(ref.owner_rider_id) === String(riderId)) throw new Error('Cannot use your own referral code');

  const newUser = await getRider(riderId);
  newUser.wallet_balance += 50.0;
  newUser.referred_by = code;
  await newUser.save();

  const owner = await getRider(ref.owner_rider_id);
  owner.wallet_balance += 25.0;
  await owner.save();

  ref.times_used += 1;
  ref.reward_earned += 25.0;
  await ref.save();

  return {
    success: true,
    new_user_bonus: 50.0,
    referrer_bonus: 25.0,
    message: '₹50 added to your wallet!',
  };
}

// ── Format helpers ────────────────────────────────────────────────────────

function formatRider(r) {
  return {
    id: r._id ? r._id.toHexString() : r.id,
    name: r.name,
    phone: r.phone,
    city: r.city,
    zone: r.zone,
    platform: r.platform,
    age: r.age,
    walletBalance: r.wallet_balance,
    isPolicyActive: r.is_policy_active,
    policyTier: r.policy_tier,
    registeredAt: r.registered_at,
    referralCode: r.referral_code,
    referredBy: r.referred_by,
    workerId: r.worker_id,
    verified: r.verified,
  };
}

function formatPayoutLog(p) {
  return {
    id: p._id ? p._id.toHexString() : p.id,
    riderId: p.rider_id,
    amount: p.amount,
    timestamp: p.timestamp,
    triggerType: p.trigger_type,
    zone: p.zone,
    claimNumber: p.claim_number,
  };
}

module.exports = {
  registerRider,
  getRider,
  updateRider,
  getPayouts,
  getQuote,
  buyPolicy,
  getClaimTracker,
  getMarketStatus,
  generateReferralCode,
  redeemReferral,
  formatRider,
  formatPayoutLog,
  isMarketCrashActive,
  setMarketCrashState,
};







